### PR TITLE
Skip tests on a docs-only PR

### PR DESCRIPTION
**Pull request by tcsc**

### CODE REVIEW

File: .cloudbuild/scripts/can-skip/main.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
+/*
```
Comment (russjones) [inline .cloudbuild/scripts/can-skip/main.go:1]: This is great. :+1:

**Comment by klizhentas** [comment]
@tcsc not sure if this should be part of the request, but old CI also ran yarn markdown-lint and validated links:

https://github.com/gravitational/teleport/blob/branch/v7/.drone.yml#L284

https://github.com/gravitational/teleport/blob/branch/v7/build.assets/Makefile#L334

**Comment by tcsc** [comment]
> not sure if this should be part of the request, but old CI also ran yarn markdown-lint and validated links:

That's handled by a separate trigger so it can run in parallel with the other build steps. See "https://github.com/gravitational/teleport/blob/master/.cloudbuild/ci/doc-tests.yaml

File: .cloudbuild/ci/integration-tests.yaml [hunk 2]
```diff
@@ -1,29 +1,69 @@
+timeout: 20m
+options:
+  machineType: E2_HIGHCPU_32
+  
+  volumes:
+    # A place to put shared state between build steps so that we don't pollute
+    # the workspace
+    - name: state
+      path: '/state'
```
Comment (russjones) [inline .cloudbuild/ci/integration-tests.yaml:9]: Hrm, how does this `/state` directory work? Can that submits a PR write to it?

File: .cloudbuild/scripts/can-skip/main.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+)
+
+// main is just a stub that prints out an error message and sets a nonzero exit
+// code on failure.All of the work happens in `innerMain()`.
+func main() {
+	if err := innerMain(); err != nil {
+		fmt.Printf("FAILED: %s\n", err.Error())
+		os.Exit(-1)
```
Comment (russjones) [inline .cloudbuild/scripts/can-skip/main.go:39]: `log.Fatalf`

File: .cloudbuild/scripts/can-skip/main.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, fmt.Errorf("failed getting commit %q worktree: %w", sha, err)
+	}
+
+	return tree, nil
+}
+
+func touch(filename string) error {
+	f, err := os.Create(filename)
+	if err != nil {
+		return fmt.Errorf("failed touching file %s: %w", filename, err)
```
Comment (russjones) [inline .cloudbuild/scripts/can-skip/main.go:182]: `trace.Wrap` here and elsewhere.

File: .cloudbuild/ci/integration-tests.yaml [hunk 5]
```diff
@@ -1,29 +1,69 @@
...
   - name: us-docker.pkg.dev/ci-account/teleport/buildbox-root:v0.1.0
     id: root-integration-test
-    args: ['make', 'rdpclient', 'integration-root']
+    entrypoint: bash
+    args: 
+      - '-c'
+      - |
+        if [ ! -f $$SKIP_TESTS ]; then 
+          make rdpclient integration-root
+        fi
```
Comment (russjones) [inline .cloudbuild/ci/integration-tests.yaml:42]: I think this is fine for now to unblock docs writers, but I think we should just execute a program like your `can-skip` as a wrapper checks if its a docs PR and running unit and integration tests can be skipped. If it not, it can just `os.exec` `make tests`.

This has several benefits. It makes each run stateless and you don't need to persist a `/state` directory.

This approach has another benefit, you can get rid of `(make run-etcd &); sleep 1; make integration` and instead launch etcd, wait for etcd to be healthy https://etcd.io/docs/v3.4/op-guide/monitoring/#health-check then run tests.

Basically let's follow a similar approach to what we are doing with the workflow automation bot, reduce as much logic in YAML, push it all to a thin Go program.

File: .cloudbuild/ci/integration-tests.yaml (hunk 2 above)
Comment (tcsc) [reply .cloudbuild/ci/integration-tests.yaml:9]: It's an ephemeral volume that gets mounted in every step and discarded after the build

File: .cloudbuild/scripts/can-skip/main.go (hunk 4 above)
Comment (tcsc) [reply .cloudbuild/scripts/can-skip/main.go:182]: No worries. Before this was a full-on module I was trying to avoid pulling in extra packages and using the built-in go 1.13+ error wrapping. This doesn't matter as much any more, as I'm already pulling in a bunch of dependencies - and consistency with the rest of the codebase is more important.

**Review by russjones — APPROVED** [review]

File: .cloudbuild/ci/integration-tests.yaml (hunk 2 above)
Comment (russjones) [reply .cloudbuild/ci/integration-tests.yaml:9]: Let's talk about this tomorrow, have to think about this some more.

File: .cloudbuild/scripts/internal/changes/changes.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+	return report, nil
+}
+
+func isDocChange(change *object.Change) bool {
+	path := getChangePath(change)
+	return path == "" ||
+		strings.HasPrefix(path, "docs/") ||
+		strings.HasSuffix(path, ".mdx") ||
+		strings.HasSuffix(path, ".md")
```
Comment (greedy52) [inline .cloudbuild/scripts/internal/changes/changes.go:78]: nit: in what case `path == ""`? and maybe also lower case `path` for comparison just to be safe.

File: .cloudbuild/scripts/internal/etcd/start.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+	"net"
+	"os"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// Start starts the etcd serve using the Makefile `run-etcd` task and waits for it to start.
```
Comment (greedy52) [inline .cloudbuild/scripts/internal/etcd/start.go:31]: nit: typo `serve`

**Review by greedy52 — APPROVED** [review]

File: .cloudbuild/scripts/internal/changes/changes.go (hunk 6 above)
Comment (tcsc) [reply .cloudbuild/scripts/internal/changes/changes.go:78]: The `path == ""` case was a bad cargo-cult copy out of the original shell script - have removed it.

File: .cloudbuild/scripts/internal/etcd/start.go (hunk 7 above)
Comment (tcsc) [reply .cloudbuild/scripts/internal/etcd/start.go]: Done

**Review by russjones — APPROVED** [review]
